### PR TITLE
PR: Print warning when using `global` in an empty namespace

### PR DESCRIPTION
**Pull request by impact27**
Depends on https://github.com/spyder-ide/spyder-kernels/pull/355

### CODE REVIEW

**Comment by ccordoba12** [comment]
@impact27, please rebase this one to see if we could include it in 2.3.0.

**Comment by impact27** [comment]
Python 2.7 is stopping this one from passing. Should we wait for python 2.7 support to be dropped?

File: spyder_kernels/console/tests/test_console_kernel.py [hunk 1]
```diff
@@ -1093,5 +1093,47 @@ def test_locals_globals_in_pdb(kernel):
...
+        p = tmpdir.join("test.py")
+        p.write(code)
+
+        # Run code in current namespace
+        client.execute("runfile(r'{}', current_namespace=True)".format(
+            to_text_string(p)))
+        msg = client.get_iopub_msg(timeout=TIMEOUT)
+        while "text" not in msg["content"]:
+            msg = client.get_iopub_msg(timeout=TIMEOUT)
+        assert "WARNING: This file contains a global statment" not in (
```
Comment (ccordoba12) [inline spyder_kernels/console/tests/test_console_kernel.py:1126]:
```suggestion
        assert "WARNING: This file contains a global statement" not in (
```

File: spyder_kernels/console/tests/test_console_kernel.py [hunk 2]
```diff
@@ -1093,5 +1093,47 @@ def test_locals_globals_in_pdb(kernel):
...
+            msg["content"]["text"])
+
+        # Run code in empty namespace
+        client.execute("runfile(r'{}')".format(to_text_string(p)))
+        msg = client.get_iopub_msg(timeout=TIMEOUT)
+        while "text" not in msg["content"]:
+            msg = client.get_iopub_msg(timeout=TIMEOUT)
+        assert "WARNING: This file contains a global statment" in (
```
Comment (ccordoba12) [inline spyder_kernels/console/tests/test_console_kernel.py:1134]:
```suggestion
        assert "WARNING: This file contains a global statement" in (
```

File: spyder_kernels/customize/spydercustomize.py [hunk 3]
```diff
@@ -471,6 +474,19 @@ def exec_code(code, filename, ns_globals, ns_locals=None, post_mortem=False,
         else:
             ast_code = ast.parse(transform_cell(code))
 
+        # Print warning for global
+        if global_warning and SHOW_GLOBAL_MSG:
+            has_global = any(
+                isinstance(node, ast.Global) for node in ast.walk(ast_code))
+            if has_global:
+                _print(
+                    "\nWARNING: This file contains a global statment, "
```
Comment (ccordoba12) [inline spyder_kernels/customize/spydercustomize.py:483]:
```suggestion
                    "\nWARNING: This file contains a global statement, "
```

File: spyder_kernels/customize/spydercustomize.py [hunk 4]
```diff
@@ -471,6 +474,19 @@ def exec_code(code, filename, ns_globals, ns_locals=None, post_mortem=False,
...
+                    "but it is run in an empty namespace. "
+                    "Consider using the "
+                    "'Run in console's namespace instead of an empty one' "
+                    "option if you want to capture the namesapce.\n")
```
Comment (ccordoba12) [inline spyder_kernels/customize/spydercustomize.py:487]:
```suggestion
                    "option, that you can find in the menu 'Run > "
                    "Configuration per file', if you want to capture the "
                    "namespace.\n")
```

**Review by ccordoba12 — COMMENTED** [review]
Good idea @impact27, thanks for this! I left some minor suggestions for you, otherwise looks good to me.

About your comment: 

> Python 2.7 is stopping this one from passing. Should we wait for python 2.7 support to be dropped?

Ok, no problem. Let's leave this improvement for Spyder 6 then.

**Comment by ccordoba12** [comment]
@impact27, this looks pretty good! Please solve the merge conflicts, so we can include this in our next version (2.3.1), which I plan to release tomorrow.

**Comment by impact27** [comment]
The PR doesn’t work with python 2.

**Comment by ccordoba12** [comment]
Ok, I see. Then we have to leave it for Spyder-kernels 3 then.

File: spyder_kernels/console/tests/test_console_kernel.py [hunk 5]
```diff
@@ -1114,5 +1114,47 @@ def test_get_interactive_backend(backend):
...
+    """
+    # Command to start the kernel
+    cmd = "from spyder_kernels.console import start; start.main()"
+
+    with setup_kernel(cmd) as client:
+        # Remove all variables
+        client.execute("%reset -f")
+        client.get_shell_msg(timeout=TIMEOUT)
+
+        # Write multiprocessing code to a file
```
Comment (ccordoba12) [inline spyder_kernels/console/tests/test_console_kernel.py:1129]:
```suggestion
        # Write code with a global to a file
```

File: spyder_kernels/console/tests/test_console_kernel.py [hunk 6]
```diff
@@ -1114,5 +1114,47 @@ def test_get_interactive_backend(backend):
...
+        code = (
+            "def foo1():\n"
+            "    global x\n"
+            "    x = 2\n"
+            "x = 1\n"
+            "print(x)\n"
+            )
```
Comment (ccordoba12) [inline spyder_kernels/console/tests/test_console_kernel.py:1136]:
```suggestion
        )
```

File: spyder_kernels/customize/spydercustomize.py [hunk 7]
```diff
@@ -393,6 +396,21 @@ def exec_code(code, filename, ns_globals, ns_locals=None, post_mortem=False,
         else:
             ast_code = ast.parse(transform_cell(code))
 
+        # Print warning for global
+        if global_warning and SHOW_GLOBAL_MSG:
+            has_global = any(
+                isinstance(node, ast.Global) for node in ast.walk(ast_code))
```
Comment (ccordoba12) [inline spyder_kernels/customize/spydercustomize.py:402]: Could this make running large files (i.e. with more than 1000 lines of code) slow?

File: spyder_kernels/customize/spydercustomize.py [hunk 8]
```diff
@@ -393,6 +396,21 @@ def exec_code(code, filename, ns_globals, ns_locals=None, post_mortem=False,
...
+            if has_global:
+                print(
+                    "\nWARNING: This file contains a global statement, "
+                    "but it is run in an empty namespace. "
+                    "Consider using the "
+                    "'Run in console's namespace instead of an empty one' "
+                    "option, that you can find in the menu 'Run > "
+                    "Configuration per file', if you want to capture the "
+                    "namespace.\n")
```
Comment (ccordoba12) [inline spyder_kernels/customize/spydercustomize.py:411]:
```suggestion
                    "namespace.\n"
                )
```

**Review by ccordoba12 — COMMENTED** [review]
Thanks @impact27 for updating this PR! I think this is almost ready.

File: spyder_kernels/customize/spydercustomize.py (hunk 7 above)
Comment (impact27) [reply spyder_kernels/customize/spydercustomize.py:402]: Slightly slower, but this shouldn't take too long. See for example:
```python
In [1]: import ast

In [2]: code = "a = 123\n" * 10000

In [3]: ast_code = ast.parse(code)

In [4]: %timeit any(isinstance(node, ast.Global) for node in ast.walk(ast_code))
47.9 ms ± 821 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Comment (impact27) [reply spyder_kernels/customize/spydercustomize.py:402]: The complexity here is linear, so I am not too worried.

Comment (ccordoba12) [reply spyder_kernels/customize/spydercustomize.py:402]: Ok, that's a really small delay. Thanks for checking it.

**Review by ccordoba12 — APPROVED** [review]
Thanks @impact27!

**Pull request merged**